### ISSUE TITLE
Add support for uploading and rendering text files

### ISSUE THREAD

**Issue by kozak-codes** [mastodon/mastodon#21643]
### Pitch

Add the ability to upload .txt or .md files and have them rendered when you click on them, similar to how images will expand in size. Users would not be able to view the rendered text until they click on the name of the file that has been uploaded. Some metadata could be added to the file, and they should show similar to external links.

Mastodon is mainly a micro blogging platform, but it's also a rich media platform - for example, we have the capability to share images, videos, polls, links, urls. We can share all of this rich content but what if we wanted to share something with a bit more content? 

Current alternatives to sharing long form content:
 - Use an alternative activitypub/federated app: write.as, frendica, diaspora, hubzilla
 - creating an actual blog, eg wordpress
 - writing 16 toots and replying to each one
 - writing it in the notes file and uploading a picture
 - Using an OEmbed tool like https://paste.social/

All of these alternatives require the user to leave the app, open a new tab, etc to read the content and IMO that isn't fair to users. I like the micro blogging format - it makes sense, and makes the timeline easier to read. But I think exploring a simple solution using the systems that we already have in place would go a long way.

Here is an example of what the toot would look like:
![image](https://user-images.githubusercontent.com/1368558/204060028-41b01c64-0078-4332-965b-7bd7a4f51d1b.png)

And this is what the markdown would look like, similar to how images are rendered:
![image](https://user-images.githubusercontent.com/1368558/204060311-ccf65131-2ff3-464a-8dbb-1d48460722c0.png)

I don't think an editor for markdown files would be necessary - it would be up to users to find their own WYSIWG editor for markdown if that's their jam. Simply allow .md and .txt files to be uploaded & attached to toots. The title & thumbnail of the file should be configurable as well while it should show the name of the file. It would be nice if the markdown supported emojis, but I don't think it's strictly necessary.

Security considerations: how secure is the markdown renderer? There would have to be zero trust from other mastodon instances when rendering the markdown and we'd have to be certain that this would not introduce any XSS vulnerabilities.

Related to: #21224 #20100 

### Motivation

This would allow users to upload long form content in a way that doesn't impede current functionality, placing the long form content as a back-burner. The user still has to "hook" others into reading their long form content through the toot. It would be no more intrusive than simply providing a link and could build on existing systems.

Users could use this to share code, write articles, make ascii art, build an FAQ about their page, direct users to their content, or make detailed lists of users to follow. All without leaving mastodon and in an unobtrusive format.

**Comment by kozak-codes** [comment]
I'd be willing to take this on if @Gargron or another dev thinks it's within the scope of the system

**Comment by robsonsobral** [comment]
I've been thinking about your idea, and it could do more harm than good. Imagine how easy would be to distribute "evil" content behind a "good" toot and how hard would be to avoid it.

**Comment by eishiya** [comment]
I don't think Mastodon is the place for long-form text content. The kinds of media that Mastodon currently supports are all essentially short-form - images, small videos, short-ish audio (due to size limitations). These *can* be used for long-form content (people can put text in images, or post long-ish audio), but it's not convenient and very much feels like a hack.

I'd much rather see long-form text content continue to be something people have to link to. There are other platforms for long-form posts that are much more convenient than Mastodon ever could be for them, and where people go specifically because they want something juicier to read than microposts.

I see no technical difficulty with supporting text files, and I can even see some short-form use cases for them (code snippets, sharing poetry with presentation more akin to artwork instead), and I think that perhaps filters could be extended to work on text files to avoid some of the abuse potential, but I don't think Mastodon should encourage users to use it as a long-form text platform. Keep that external.

**Comment by ansiwen** [comment]
@robsonsobral 
> I've been thinking about your idea, and it could do more harm than good. Imagine how easy would be to distribute "evil" content behind a "good" toot and how hard would be to avoid it.

I can't follow that argument. How is that a specific problem of Markdown attachment, and how does it make any situation worse? I only see an improvement in usability and no increased abusability. What you describe here can be done already with external links and their previews or other attachments, where it's even more difficult to detect. The difference is only, that you can use mastodons file storage, and you don't need to use external web servers.

Apart from that, there is software on the fediverse, like https://writefreely.org which is specialized on this usage. So, not allowing it in Mastodon makes it more difficult for valid use-cases but does not keep away abusers who can just use their own publisher.

**Comment by ansiwen** [comment]
@eishiya 
> I'd much rather see long-form text content continue to be something people have to link to. 

Isn't that exactly what this Issue is proposing? A possibility to use long-text as attachments and link to it?

Inline long-text in the post itself is possible and common already (the popular glitch fork supports arbitrarily long posts and markdown syntax). And there is demand for it, which you can see from the obnoxious "threads" practice from twitter.

So, I see this proposal exactly as a _fix_ for what you are arguing for, to keep the long text out of the actual post.

**Comment by eishiya** [comment]
> Isn't that exactly what this Issue is proposing? A possibility to use long-text as attachments and link to it?

By link, I meant external link. Which is what people do now if their blog isn't in the Fediverse. Including the text as media is not a link, it's part of the post.

Mastodon is not a long-format platform and we should not encourage people to use it as such. People can already follow federated long-form blogging (Plume, etc) if they want long-form text directly on their timeline, or join such an instance if they're looking to post such content.

Maybe it's just a difference between the instances we're on, but IME, threads on Mastodon tend to be streams of consciousness that users add on to after they think they're done, and the formal Thread:tm: of Twitter where people try to duplicate what is essentially a long-form blog post is rare and not particularly welcome, precisely because there are better alternatives that still let you reach the same audience, unlike Twitter where getting people to follow you elsewhere is nigh impossible.

**Comment by robsonsobral** [comment]
Emphasis are mine:

> What you describe here can be done already with **external links and their previews or other attachments, where it's even more difficult to detect**. The difference is only, that you can use mastodons file storage, and you don't need to use external web servers.

Yes, @ansiwen , it's harder to detect already. Why make it easier to use? I don't see long texts as a problem so big that it should be replaced by another. As you obviously do hate long texts, can't you find a different solution?

**Comment by kozak-codes** [comment]
Security and accesibility are huge benefits of this format.

This is more secure compared to a link to an external site because the text is immutable - you'll have the same experience as someone else that read the post 2 years ago. There are less security concerns because you arent being tracked by some other website (or worse - potential exposure to malware)

When you post a markdown file (or an image with words) you need to post a 500 character message explaining why the person wants to read it. It is completely opt in and if the "hook" is different than the content then you can always block that individual or unfollow them. At the very least, you havent clicked on a website with 60 ads and 7 trackers.

Accessibility is critical for any large application these days and something that twitter didnt really do very well. Recently I saw a post that had an image attached with a picture of their notes app. It was a really great account of someone's experience in a war torn country that was very valuable however someone that used a screenreader woupd not have been able to read that account. This format is troublesome because it can't easily be read by a screenreader.

Do we have any other good alternative long form solutions for people that want to use mastodon/acitivity pub in a secure way that maintains a consistent "stream of thought" thats also just as accessible? People currently try and find workarounds and IMO this is a very simple, elegant solution that would bring so much value to the existing experience without sacrificing anything.

**Comment by robsonsobral** [comment]
> People currently try and find workarounds and IMO this is a very simple, elegant solution that would bring so much value to the existing experience without sacrificing anything.

To me, it's a workaround. A workaround which can cause a different harm than the one you're trying to fix.

In any case, it's a huge change. Easy to implement, but giant to the use of the platform. It could change how people interact with it entirely. So, it requires careful thoughts.

**Comment by ansiwen** [comment]
@eishiya 
> By link, I meant external link. Which is what people do now if their blog isn't in the Fediverse. Including the text as media is not a link, it's part of the post.

It could be displayed exactly in the same way like previews to websites. So I don't see how that makes a difference if it is an external website or a file in the media folder.

> Mastodon is not a long-format platform and we should not encourage people to use it as such. People can already follow federated long-form blogging (Plume, etc) if they want long-form text directly on their timeline, or join such an instance if they're looking to post such content.
> 
> Maybe it's just a difference between the instances we're on, but IME, threads on Mastodon tend to be streams of consciousness that users add on to after they think they're done, and the formal Thread™️ of Twitter where people try to duplicate what is essentially a long-form blog post is rare and not particularly welcome, precisely because there are better alternatives that still let you reach the same audience, unlike Twitter where getting people to follow you elsewhere is nigh impossible.

Well, whenever you see a post which ends with something like `1/22`, it is obviously planned as long-post from the beginning, and I see that very often unfortunately, and with the Twitter exodus more of that will come. You seem to assume that "not encouraging" people to do long-posts by making it more uncomfortable would lead to people use proper platforms in the fediverse or use external links. But that completely contradicts my experience. People take the path of least resistance, and that is not creating an external resource and link to it, but to make long threads. It's unfortunate, but it's how it is. But if there is an _easy_ way to create these "external links", then you can tell these long-thread-posters: "why you don't just create a proper text attachment?" and it will be much more likely to be adopted.

**Comment by ansiwen** [comment]
> Emphasis are mine:
> 
> > What you describe here can be done already with **external links and their previews or other attachments, where it's even more difficult to detect**. The difference is only, that you can use mastodons file storage, and you don't need to use external web servers.
> 
> Yes, @ansiwen , it's harder to detect already. Why make it easier to use? I don't see long texts as a problem so big that it should be replaced by another. As you obviously do hate long texts, can't you find a different solution?

No, I don't hate long texts, quite the opposite. And let's please avoid ad hominem statements and stay focused on the topic.

This issue proposes a solution to _allow_ longer texts, while mastodon is limited to 500 chars by default. And there is a demand among mastodon users for writing longer texts as a matter of fact. We can let people work around that by "abusing" the short message format and chaining many posts together, as they are doing now, or we can lead them to something better. One of the better options would be text upload as described in this issue. Another one would be to allow long messages by default and add a "see more" button after 500 chars. And probably there are other options. But "use another software" or "link to external blogs" will not work for these users, they will instead continue to write long threads.

So, it there is a problem, and a solution. If you see other problems coming with this solution, can you elaborate on them? Because I see the same possibilities for filtering out unwanted content as you can do now.

**Comment by robsonsobral** [comment]
> Because I see the same possibilities for filtering out unwanted content as you can do now.

No, you can't filter unwanted content the same way you do today.


> We can let people work around that by "abusing" the short message format and chaining many posts together, as they are doing now, or we can lead them to something better.

Actually, people would work around limitations and take advantage of them. For example, there are many artists using the possibility of chaining posts. They turned the limitation into something new. Quoting myself:

> [...] it's a huge change. Easy to implement, but giant to the use of the platform. It could change how people interact with it entirely. So, it requires careful thoughts.

A "see more" button is a much less drastic change than to create an Instagram of slides. Even better, I would like to have less friction between the items of a thread.

It's a big change!

**Comment by kozak-codes** [comment]
The primary difference between this proposal and "read more" would be the addition of markdown to text. The ability to add quotes, hyperlinks, emphasis/underlines, lists, checkboxes, and **perhaps** images which can all serve many different purposes. Putting these on the timeline would be too much clutter.

**Comment by JiFish** [comment]
It's a huge shame this issue never got any traction.

I recently posted some d&d monster stat blocks. I settled on posting them as images, with a copy of the text in image description. Since you get 4 images with 1,500 characters of text on each... it's crazy to me that I couldn't just attach a small amount of text and save myself a lot of trouble.

IMO, this is no different from a toot that says "hey check out this track I wrote and played" and attaching a flac file. I'm not trying to use mastodon as a long-form blogging site, I just occasionally have the need to attach a larger amount of text separate from the main toot.

**Comment by rauschma** [comment]
I’m regularly seeing screenshots of text on Mastodon, with plain text descriptions. In a way, the image it useless information in this case.

**Comment by brendanjones** [comment]
I'd like to be able to attach files to posts, but I don't need them to render. My computer renders files for me already, y'know? With other apps. Rendering feels like a massive, unnecessary feature addition with implications already covered by other commenters above. 

The attachment-only part of this request is covered more generically by https://github.com/mastodon/mastodon/issues/26420 (because if attachments are allowed, why limit to .txt and .md?)

**Comment by kevATin** [comment]
> I’m regularly seeing screenshots of text on Mastodon, with plain text descriptions. In a way, the image it useless information in this case.

Not to mention that those text-images are probably (i would presume) useless to people with screen readers. 


I don't see how features that allow for more and fancier text are always so controversial for Mastodon. 
If you don't want to look at the whole thing, you can just scroll past it, just like with an image or video. 
And even if everyone started publishing their books on Mastodon, i don't think the storage space requirements would come close to the video ones.
